### DISCUSSION TITLE
[FeatureRequest] Transforms: Add Script Option to add field from calculation transform

### DISCUSSION THREAD

**Discussion by aldoferrante** [grafana/grafana#38195] [Dashboards, Panels & Visualizations]
**What would you like to be added:**
Add a script option in the visualization panel transform tab mode select.

**Why is this needed**:
Currently only 2 options are available in the panel transform tab, "Binary Operation" & "Reduce Row", though useful, it may not provide flexibility for custom transformations. A JavaScript option provides more flexibility in transforming the data for application specific requirements. For example, storing a JSON in Influx DB string field, using JavaScript transformation in a visualization panel, we can parse the data from the JSON to display or reorder data and much much more.  A transform JavaScript option would address many customization requirements.
![image](https://user-images.githubusercontent.com/31166374/103019721-42116c80-4515-11eb-919d-a5c757794f5c.png)

**Comment by aldoferrante** [comment]
Additional comment to illustrate the need. We created a custom table to calculate a KPI and parse the JSON to filter certain variables. We used an SVG panel and wrote JavaScript in addition to SVG code to visualize the data. A better approach would be to use the standard Gauge Panel and write a JavaScript in the transform tab and not have to write code for visualization.

**Reply by koen-lee** [reply]
Another use case: with temperature and relative humidity, one could compute an absolute humidity and dew point, useful to alert on condensation risk or to decide whether ventilation will dehumidify.

Also useful:  to create a proper average for wind direction (in degrees), a way is to split it in the sine and cosine and scale by wind speed. This prevents almost-north wind, 359 and 1 degrees, to average to south wind (180 degrees) which is obviously wrong.

**Comment by pkolyvas** [comment]
We're in the process of moving Grafana feature requests to [GitHub Discussions](https://community.grafana.com/t/were-moving-grafana-feature-requests-to-github-discussions/50341/3). This issue is being moved to [Discussions](https://github.com/grafana/grafana/discussions) or has been moved from our issues list, depending on where you're seeing this message. 

 We hope to use the discussion thread to continue building use cases, and additional evidence based on engagement/voting, in order to help us prioritize this feature request later on down the road. 

As with all things Grafana, we value your input and participation in this project and appreciate your patience during this transitional period. Looking forward to the discussion!

**Comment by palsivertsen** [comment]
I have a similar use case where I want to visualize individual bits in one or more bitmaps. Having an option where I could transform rows using a JavaScript function would be really useful.

I'm picturing a transformation that would let me define a JavaScript function that takes a `row` and returns the transformed `row`:

```javascript
function(row) {
  row.myValue = Boolean(row.myBitmap & 8) // checks if the 4th bit is set  (8 = 0000 1000)
  row.sum = row.a + row.b
  return row
}
```

The above function would transform a row from:
```json
{
  "myBitmap": 10,
  "a": 1,
  "b": 2
}
```

To:
```json
{
  "myBitmap": 10,
  "a": 1,
  "b": 2,
  "myValue": true,
  "sum": 3
}
```

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). I believe this feature request is captured here https://github.com/grafana/grafana/issues/33154 . Feel free to make an [issue](https://github.com/grafana/grafana/issues) if you cannot find a solution and would like to request the feature.

**Comment by adminy** [comment]
Um, it feels like this still hasn't been addressed. It sounds like a vital feature tbh, I have a query coming from cloudwatch insights, which of course it has to come pre-grouped by a field. So naturally, I have it grouped to more than 1 field so that I can get the data out.

First I find that grafana hasn't got a url type when converting fields, nor does it have any sort of string manipulation field transformation methods. 

I would be happy with either a 
```javascript
function(row) { return row }
// OR
function(column) { return column }
```
transformation.
